src/shell/shell_cmds.cpp

```diff
@@ -465,26 +472,250 @@ static std::string to_search_pattern(const char *arg)
+			else if(optAminusD && (result.attr&DOS_ATTR_DIRECTORY) ) continue;
+			else if(optAS && !(result.attr&DOS_ATTR_SYSTEM) ) continue;
+			else if(optAminusS && (result.attr&DOS_ATTR_SYSTEM) ) continue;
+			else if(optAH && !(result.attr&DOS_ATTR_HIDDEN) ) continue;
+			else if(optAminusH && (result.attr&DOS_ATTR_HIDDEN) ) continue;
+			else if(optAR && !(result.attr&DOS_ATTR_READ_ONLY) ) continue;
+			else if(optAminusR && (result.attr&DOS_ATTR_READ_ONLY) ) continue;
+			else if(optAA && !(result.attr&DOS_ATTR_ARCHIVE) ) continue;
+			else if(optAminusA && (result.attr&DOS_ATTR_ARCHIVE) ) continue;
+			else if(!(optA||optAD||optAminusD||optAS||optAminusS||optAH||optAminusH||optAR||optAminusR||optAA||optAminusA) && (result.attr&DOS_ATTR_SYSTEM || result.attr&DOS_ATTR_HIDDEN) && strcmp(result.name, "..") ) continue;
+
+			results.push_back(result);
+
+		} while ( (ret=DOS_FindNext()) );
+		lfn_filefind_handle=fbak;
+
+		if (optON) {
+			// Sort by name
+			std::sort(results.begin(), results.end(), DtaResult::compareName);
+		} else if (optOE) {
+			// Sort by extension
+			std::sort(results.begin(), results.end(), DtaResult::compareExt);
+		} else if (optOD) {
+			// Sort by date
+			std::sort(results.begin(), results.end(), DtaResult::compareDate);
+		} else if (optOS) {
+			// Sort by size
+			std::sort(results.begin(), results.end(), DtaResult::compareSize);
+		} else if (optOG) {
+			// Directories first, then files
+			std::sort(results.begin(), results.end(), DtaResult::groupDirs);
+		} else if (optO) {
+			// Directories first, then files, both sort by name
+			std::sort(results.begin(), results.end(), DtaResult::groupDef);
+		}
+		if (reverseSort) {
+			std::reverse(results.begin(), results.end());
+		}
+
+		for (std::vector<DtaResult>::iterator iter = results.begin(); iter != results.end(); ++iter) {
+
+			char * name = iter->name;
+			char *lname = iter->lname;
+			Bit32u size = iter->size;
+			Bit16u date = iter->date;
+			Bit16u time = iter->time;
+			Bit8u attr = iter->attr;
+
+			/* output the file */
+			if (optB) {
+				// this overrides pretty much everything
+				if (strcmp(".",uselfn&&!optZ?lname:name) && strcmp("..",uselfn&&!optZ?lname:name)) {
+					shell->WriteOut("%s\n",uselfn&&!optZ?lname:name);
+				} else {
+					// skip to the next file, otherwise this file
+					// will be counted as printed for pause cmd
+					continue;
+				}
+			} else {
+				if (first&&optS) {
+					first=false;
+					shell->WriteOut("\n");
+					shell->WriteOut(MSG_Get("SHELL_CMD_DIR_INTRO"),uselfn&&!optZ&&DOS_GetSFNPath(path,largs,true)?largs:sargs);
+					if (optP) {
+						p_count+=optW?10:2;
+						if (p_count%(GetPauseCount()*w_size)<3) {
+							shell->WriteOut(MSG_Get("SHELL_CMD_PAUSE"));
+							Bit8u c;Bit16u n=1;
+							DOS_ReadFile(STDIN,&c,&n);
+							if (c==3) {shell->WriteOut("^C\r\n");return false;}
+							if (c==0) DOS_ReadFile(STDIN,&c,&n); // read extended key
+							shell->WriteOut_NoParsing("\n");
+						}
```

Comment:
(Here's another copy & paste'd chunk of code from above)

Reply:
This has now been lifted out to a sub-function.